Fix issues with the Win32 implementation of the POSIX dynamic library API

spencercw opened the pull request:
dlerror() is supposed to return a char *, but currently returns a DWORD. Reimplement it using the Win32 FormatMessage function.

The only place dlerror() is used is ssh-pkcs11.c, which is not currently compiled into the Windows version. However, PR #331 enables PKCS#11 support, so if that PR is merged then this one should be also.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/PowerShell/openssh-portable?pullRequest=379) <br/>All CLA requirements met.

manojampalam commented on contrib/win32/win32compat/misc.c:
```diff
@@ -237,9 +242,33 @@ dlclose(HMODULE handle)
+	if (last_dlerror == ERROR_SUCCESS)
+		return NULL;
+
+	DWORD length = FormatMessageA(
```
Can you please double check if FormatMessageA works for Unicode strings and spits out UTF-8 output? 

If that's not the case, I believe you should be using FormatMessageW followed by utf16_to_utf8.

spencercw replied:
This will presumably use the current code page rather than UTF-8. I will change it to use FormatMessageW.

manojampalam commented on contrib/win32/win32compat/misc.c:
```diff
@@ -237,9 +242,33 @@ dlclose(HMODULE handle)
+	if (length == 0)
+		return "Failed to format error message";
+
+	return message;
```
What's the expectation on how "message" is freed? Looks like a simple free() on this can lead to issues as its supposed to be LocalFree()d

If this is true, you will have to return and xcopy version and LocalFree original version.

spencercw replied:
Per the [spec](http://pubs.opengroup.org/onlinepubs/9699919799/functions/dlerror.html): The returned pointer might be invalidated or the string content might be overwritten by a subsequent call to dlerror() in the same thread (if dlerror() is thread-safe) or in any thread (if dlerror() is not thread-safe). The returned pointer might also be invalidated if the calling thread is terminated.

In other words, the user does not free the message. This is why there is a LocalFree() at the top of the function.

spencercw commented:
I have changed it to trim the trailing line break as well to be more conformant to the dlerror() spec.

spencercw committed:
Fix Win32 implementation of dlerror()

dlerror() is supposed to return a char *, but currently returns a
DWORD. Reimplement it using the Win32 FormatMessage function.

spencercw commented:
Fixed a warning..

manojampalam commented on contrib/win32/win32compat/misc.c:
```diff
@@ -221,10 +221,15 @@ explicit_bzero(void *b, size_t len)
 dlopen(const char *filename, int flags)
 {
-	return LoadLibraryA(filename);
+	HMODULE module = LoadLibraryA(filename);
```
Sorry, missed this in earlier review.. You will have to convert filename to utf16 and call LoadLibraryW, to have this support Unicode file paths.

spencercw committed:
Correctly handle UTF-8 filenames in the Win32 implementation of dlopen()
